File: demo/demo/report/voucher_action_report/voucher_action_report.js
frappe.query_reports["Voucher Action Report"] = {
    "filters": [
        {
            "fieldname":"from_date",
            "label": __("From Date"),
            "fieldtype": "Date"
        },
        {
            "fieldname":"to_date",
            "label": __("To Date"),
            "fieldtype": "Date"
        },
        {
            "fieldname":"voucher_type",
            "label": __("Voucher Type"),
            "fieldtype": "Select",
            "options": ["Sales Order", "Sales Invoice", "Delivery Note", "Purchase Order", "Purchase Invoice", "Purchase Receipt"] 
        },
		{
			"fieldname":"status",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": [ "","Draft","To Deliver and Bill","To Bill","To Deliver","Completed","Cancelled","Overdue","To Receive and Bill", "To Receive", "To Bill", "Closed"]
		},
        {
            "fieldname":"customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "depends_on": "eval:in_list(['Sales Order','Sales Invoice','Delivery Note'], frappe.query_report.get_filter_value('voucher_type'))"
        },
        {
            "fieldname":"customer_group",
            "label": __("Customer Group"),
            "fieldtype": "Link",
            "options": "Customer Group",
            "depends_on": "eval:in_list(['Sales Order','Sales Invoice','Delivery Note'], frappe.query_report.get_filter_value('voucher_type'))"
        },
        {
            "fieldname":"supplier",
            "label": __("Supplier"),
            "fieldtype": "Link",
            "options": "Supplier",
            "depends_on": "eval:in_list(['Purchase Order','Purchase Invoice','Purchase Receipt'], frappe.query_report.get_filter_value('voucher_type'))"
        },
        {
            "fieldname":"supplier_group",
            "label": __("Supplier Group"),
            "fieldtype": "Link",
            "options": "Supplier Group",
            "depends_on": "eval:in_list(['Purchase Order','Purchase Invoice','Purchase Receipt'], frappe.query_report.get_filter_value('voucher_type'))"
        }
    ],

    
    onload: function (report) {
        report.page.add_inner_button(__('Create Combined Sales Invoice'), function () {
            console.log("Sales Invoice Button Clicked");
            let selected = report.get_checked_items();

            if (!selected || !selected.length) {
                frappe.msgprint("Please select at least one Sales Order.");
                return;
            }

            let invalid = selected.some(r => r.voucher_type !== "Sales Order");
            if (invalid) {
                frappe.msgprint("You can only create a combined Sales Invoice for Sales Orders.");
                return;
            }

            let sales_orders = selected.map(row => row.voucher_no);
            console.log("Selected Sales Orders:", sales_orders);

            frappe.call({
                method: "demo.demo.report.voucher_action_report.voucher_action_report.create_combined_sales_invoice",
                args: { sales_orders: JSON.stringify(sales_orders) },
                freeze: true,
                freeze_message: __("Creating combined Sales Invoice..."),
                callback: function (r) {
                    if (!r.exc && r.message) {
                        frappe.msgprint(__('Sales Invoice created: {0}', [r.message]));
                        frappe.set_route("Form", "Sales Invoice", r.message);
                    } else {
                        frappe.msgprint(__('Failed to create combined Sales Invoice.'));
                    }
                }
            });
        });

        report.page.add_inner_button(__('Create Combined Sales Payment Entry'), function () {
            console.log("Sales Payment Button Clicked");
            let selected = report.get_checked_items();

            if (!selected || !selected.length) {
                frappe.msgprint("Please select at least one Sales Invoice.");
                return;
            }

            let invalid = selected.some(r => r.voucher_type !== "Sales Invoice");
            if (invalid) {
                frappe.msgprint("You can only create a combined Payment Entry for Sales Invoices.");
                return;
            }

            let invoices = selected.map(row => row.voucher_no);

            frappe.call({
                method: "demo.demo.report.voucher_action_report.voucher_action_report.create_combined_payment_entry",
                args: { sales_invoices: JSON.stringify(invoices) },
                freeze: true,
                freeze_message: __("Creating combined Payment Entry..."),
                callback: function (r) {
                    if (!r.exc && r.message) {
                        frappe.msgprint(__('Payment Entry created: {0}', [r.message]));
                        frappe.set_route("Form", "Payment Entry", r.message);
                    }
                }
            });
        });

        report.page.add_inner_button(__('Create Combined Purchase Invoice'), function () {
            console.log("Purchase Invoice Button Clicked");
            let selected = report.get_checked_items();

            if (!selected || !selected.length) {
                frappe.msgprint("Please select at least one Purchase Order.");
                return;
            }

            let invalid = selected.some(r => r.voucher_type !== "Purchase Order");
            if (invalid) {
                frappe.msgprint("You can only create a combined Purchase Invoice for Purchase Orders.");
                return;
            }

            let purchase_orders = selected.map(row => row.voucher_no);
            console.log("Selected Purchase Orders:", purchase_orders);

            frappe.call({
                method: "demo.demo.report.voucher_action_report.voucher_action_report.create_combined_purchase_invoice",
                args: { purchase_orders: JSON.stringify(purchase_orders) },
                freeze: true,
                freeze_message: __("Creating combined Purchase Invoice..."),
                callback: function (r) {
                    if (!r.exc && r.message) {
                        frappe.msgprint(__('Purchase Invoice created: {0}', [r.message]));
                        frappe.set_route("Form", "Purchase Invoice", r.message);
                    } else {
                        frappe.msgprint(__('Failed to create combined Purchase Invoice.'));
                    }
                }
            });
        });

        report.page.add_inner_button(__('Create Combined Purchase Payment Entry'), function () {
            console.log("Purchase Payment Button Clicked");
            let selected = report.get_checked_items();

            if (!selected || !selected.length) {
                frappe.msgprint("Please select at least one Purchase Invoice.");
                return;
            }

            let invalid = selected.some(r => r.voucher_type !== "Purchase Invoice");
            if (invalid) {
                frappe.msgprint("You can only create a combined Payment Entry for Purchase Invoices.");
                return;
            }

            let purchase_invoices = selected.map(row => row.voucher_no);
            console.log("Selected Purchase Invoices:", purchase_invoices);

            frappe.call({
                method: "demo.demo.report.voucher_action_report.voucher_action_report.create_combined_purchase_payment_entry",
                args: { purchase_invoices: JSON.stringify(purchase_invoices) },
                freeze: true,
                freeze_message: __("Creating combined Purchase Payment Entry..."),
                callback: function (r) {
                    if (!r.exc && r.message) {
                        frappe.msgprint(__('Payment Entry created: {0}', [r.message]));
                        frappe.set_route("Form", "Payment Entry", r.message);
                    } else {
                        frappe.msgprint(__('Failed to create combined Payment Entry.'));
                    }
                }
            });
        });
    },

    get_datatable_options(options) {
        options.checkboxColumn = true;
        return options;
    },

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "actions") {
            let voucher_no = data.voucher_no;
            let voucher_type = data.voucher_type;
            let update_btn = `<button class="btn btn-xs btn-secondary" onclick="custom_update_voucher('${voucher_no}', '${voucher_type}')">Update</button>`;
            let create_btn = `<button class="btn btn-xs btn-success" onclick="create_voucher('${voucher_no}', '${voucher_type}')">Create</button>`;
			return `${update_btn} ${create_btn}`;
        }

        if(column.fieldname === "view_items") {
            return `<button class="btn btn-xs btn-info" onclick="view_items_dialog('${data.voucher_type}', '${data.voucher_no}')">View</button>`;
        }

        return value;
    }
};

function custom_update_voucher(voucher_no, voucher_type) {
    if (voucher_type === "Sales Order" || voucher_type === "Purchase Order") {
        let child_doctype = voucher_type === "Sales Order" ? "Sales Order Item" : "Purchase Order Item";
        let update_method = voucher_type === "Sales Order"
            ? "demo.demo.report.voucher_action_report.voucher_action_report.update_sales_purchase_order_items_custom"
            : "demo.demo.report.voucher_action_report.voucher_action_report.update_sales_purchase_order_items_custom";

        $.when(
            frappe.call({
                method: "frappe.client.get",
                args: { doctype: voucher_type, name: voucher_no }
            }),
            frappe.db.get_doc("DocType", child_doctype)
        ).then(function(voucher_res, child_doc) {
            let voucher = voucher_res[0].message;
            let fields = [];

            child_doc.fields.forEach(function(df) {
                if (["parent", "idx", "docstatus"].indexOf(df.fieldname) === -1) {
                    fields.push({
                        fieldtype: df.fieldtype,
                        fieldname: df.fieldname,
                        label: df.label,
                        options: df.options,
                        read_only: df.read_only,
                        reqd: df.reqd,
                        precision: df.precision,
                        in_list_view: df.in_list_view
                    });
                }
            });

            fields.push({ fieldtype: 'Data', fieldname: 'docname', label: 'Docname', hidden: 1 });

            let items_data = voucher.items.map(function(item) {
                let obj = {};
                fields.forEach(function(field) {
                    obj[field.fieldname] = item[field.fieldname];
                });
                obj["docname"] = item.name;
                return obj;
            });

            let dialog = new frappe.ui.Dialog({
                title: __("Update Items for " + voucher_no),
                fields: [
                    {
                        fieldtype: 'Table',
                        fieldname: 'items',
                        label: __("Items"),
                        data: items_data,
                        fields: fields,
                        cannot_add_rows: false,
                        in_place_edit: true
                    }
                ],
                primary_action_label: __("Update"),
                primary_action(values) {
                    frappe.call({
                        method: update_method,
                        args: {
                            parent_doctype: voucher_type,
                            trans_items: JSON.stringify(values.items),
                            parent_doctype_name: voucher_no
                        },
                        callback: function(resp) {
                            if (!resp.exc) {
                                frappe.msgprint(__(voucher_type + " items updated."));
                                dialog.hide();
                                frappe.query_report.refresh();
                            } else {
                                frappe.msgprint(__("Items update failed."));
                            }
                        }
                    });
                }
            });
            dialog.show();
        });
    } else {
        frappe.set_route("Form", voucher_type, voucher_no);
    }
}

function view_items_dialog(voucher_type, voucher_no) {
    frappe.call({
        method: "demo.demo.report.voucher_action_report.voucher_action_report.get_items",
        args: {voucher_type, voucher_no},
        callback: function(r) {
            if(r.message) {
                let dialog = new frappe.ui.Dialog({
                    title: __('Items of {0}', [voucher_no]),
                    fields: [
                        {fieldtype: 'HTML', fieldname: 'items_html'}
                    ]
                });

                let html = `<table class="table table-bordered"><thead><tr>`;

                if (voucher_type === "Sales Order") {
                    html += `
                        <th>Item Code</th>
                        <th>Item Name</th>
                        <th>Total Qty</th>
                        <th>Rate</th>
                        <th>Amount</th>
                        <th>Delivered Qty</th>
                        <th>Pending Qty</th>`;
                } else {
                    html += `
                        <th>Item Code</th>
                        <th>Item Name</th>
                        <th>Qty</th>
                        <th>Rate</th>
                        <th>Amount</th>`;
                }

                html += `</tr></thead><tbody>`;

                r.message.forEach(item => {
                    html += `<tr>
                        <td>${item.item_code}</td>
                        <td>${item.item_name}</td>`;

                    if (voucher_type === "Sales Order") {
                        html += `
                            <td>${item.total_qty}</td>
                            <td>${item.rate}</td>
                            <td>${item.amount}</td>
                            <td>${item.delivered_qty || 0}</td>
                            <td>${item.pending_qty || (item.total_qty - (item.delivered_qty || 0)) || 0}</td>`;
                    } else {
                        html += `
                            <td>${item.qty}</td>
                            <td>${item.rate}</td>
                            <td>${item.amount}</td>`;
                    }

                    html += `</tr>`;
                });

                html += `</tbody></table>`;
                dialog.fields_dict.items_html.$wrapper.html(html);
                dialog.show();
            }
        }
    });
}
// Sales
function create_sales_invoice_from_so(voucher_no) {
    frappe.call({
        method: "erpnext.selling.doctype.sales_order.sales_order.make_sales_invoice",
        args: {
            source_name: voucher_no
        },
        callback: function(r) {
            if(r.message) {
                frappe.model.with_doctype("Sales Invoice", function() {
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Sales Invoice", doc.name); 
                });
            }
        }
    });
}

function create_delivery_note_from_so(voucher_no) {
    frappe.call({
        method: "erpnext.selling.doctype.sales_order.sales_order.make_delivery_note",
        args: {
            source_name: voucher_no
        },
        callback: function(r) {
            if(r.message) {
                frappe.model.with_doctype("Delivery Note", function() {
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Delivery Note", doc.name);
                });
            }
        }
    });
}

function create_payment_entry_from_si(voucher_no) {
    frappe.call({
		method: "erpnext.accounts.doctype.payment_entry.payment_entry.get_payment_entry",
        args: { dt: "Sales Invoice", dn: voucher_no },
        callback: function(r) {
            if(r.message) {
                frappe.model.with_doctype("Payment Entry", function() {
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Payment Entry", doc.name);
                });
            }
        }
    });
}

function create_delivery_note_from_si(voucher_no) {
    frappe.call({
        method: "erpnext.accounts.doctype.sales_invoice.sales_invoice.make_delivery_note",
        args: { source_name: voucher_no },
        callback: function(r) {
            if(r.message) {
                frappe.model.with_doctype("Delivery Note", function() {
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Delivery Note", doc.name);
                });
            }
        }
    });
}

function create_sales_invoice_from_dn(voucher_no) {
    frappe.call({
        method: "erpnext.stock.doctype.delivery_note.delivery_note.make_sales_invoice",
        args: { source_name: voucher_no },
        callback: function(r) {
            if(r.message) {
                frappe.model.with_doctype("Sales Invoice", function() {
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Sales Invoice", doc.name);
                });
            }
        }
    });
}
// Purchase
function create_purchase_invoice_from_po(voucher_no){
    frappe.call({
        method: "erpnext.buying.doctype.purchase_order.purchase_order.make_purchase_invoice",
        args: { source_name: voucher_no},
        callback: function(r){
            if (r.message){
                frappe.model.with_doctype("Purchase Invoice", function(){
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Purchase Invoice", doc.name);
                });
            }
        }
    });
}

function create_purchase_receipt_from_po(voucher_no){
    frappe.call({
        method: "erpnext.buying.doctype.purchase_order.purchase_order.make_purchase_receipt",
        args: { source_name: voucher_no},
        callback: function(r){
            if (r.message){
                frappe.model.with_doctype("Purchase Receipt", function(){
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Purchase Receipt", doc.name);
                });
            }
        }
    });
}

function create_purchase_receipt_from_pi(voucher_no) {
    frappe.call({
        method: "erpnext.accounts.doctype.purchase_invoice.purchase_invoice.make_purchase_receipt",
        args: { source_name: voucher_no },
        callback: function(r) {
            if (r.message) {
                frappe.model.with_doctype("Purchase Receipt", function() {
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Purchase Receipt", doc.name);
                });
            }
        }
    });
}

function create_payment_entry_from_pi(voucher_no){
    frappe.call({
        method: "erpnext.accounts.doctype.payment_entry.payment_entry.get_payment_entry",
        args:{ dt: "Purchase Invoice", dn: voucher_no},
        callback: function(r){
            if(r.message){
                frappe.model.with_doctype("Payment Entry", function(){
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Payment Entry", doc.name);
                });
            }
        }
    });
}

function create_purchase_invoice_from_pr(voucher_no){
    frappe.call({
        method: "erpnext.stock.doctype.purchase_receipt.purchase_receipt.make_purchase_invoice",
        args:{source_name: voucher_no},
        callback: function(r){
            if(r.message){
                frappe.model.with_doctype("Purchase Invoice", function(){
                    var doc = frappe.model.sync(r.message)[0];
                    frappe.set_route("Form", "Purchase Invoice", doc.name);
                });
            }
        }
    });
}

function create_voucher(voucher_no, voucher_type) {
    let dialog;
    if (voucher_type === "Sales Order") {
        dialog = new frappe.ui.Dialog({
            title: __("Create for Sales Order"),
            fields: [
                {
                    fieldtype: "Button",
                    label: __("Create Sales Invoice"),
                    click() {
                        create_sales_invoice_from_so(voucher_no);
                        dialog.hide();
                    }
                },
                {
                    fieldtype: "Button",
                    label: __("Create Delivery Note"),
                    click() {
                        create_delivery_note_from_so(voucher_no);
                        dialog.hide();
                    }
                }
            ]
        });
    } else if (voucher_type === "Sales Invoice") {
        dialog = new frappe.ui.Dialog({
            title: __("Create for Sales Invoice"),
            fields: [
                {
                    fieldtype: "Button",
                    label: __("Create Payment Entry"),
                    click() {
                        create_payment_entry_from_si(voucher_no);
                        dialog.hide();
                    }
                },
                {
                    fieldtype: "Button",
                    label: __("Create Delivery Note"),
                    click() {
                        create_delivery_note_from_si(voucher_no);
                        dialog.hide();
                    }
                }
            ]
        });
    } else if (voucher_type === "Delivery Note") {
        dialog = new frappe.ui.Dialog({
            title: __("Create for Delivery Note"),
            fields: [
                {
                    fieldtype: "Button",
                    label: __("Create Sales Invoice"),
                    click() {
                        create_sales_invoice_from_dn(voucher_no);
                        dialog.hide();
                    }
                }
            ]
        });
    } else if (voucher_type === "Purchase Order"){
        dialog = new frappe.ui.Dialog({
            title: __("Create for Purchase Order"),
            fields: [
                {
                    fieldtype:"Button",
                    label: __("Create Purchase Invoice"),
                    click(){
                        create_purchase_invoice_from_po(voucher_no);
                        dialog.hide();
                    }
                },
                {
                    fieldtype:"Button",
                    label: __("Create Purchase Receipt"),
                    click(){
                        create_purchase_receipt_from_po(voucher_no);
                        dialog.hide();
                    }
                }
            ]
        });
    } else if (voucher_type === "Purchase Invoice"){
        dialog = new frappe.ui.Dialog({
            title: __("Create for Purchase Invoice"),
            fields: [
                {
                    fieldtype: "Button",
                    label: __("Create Purchase Receipt"),
                    click(){
                        create_purchase_receipt_from_pi(voucher_no);
                        dialog.hide();
                    }
                },
                {
                    fieldtype: "Button",
                    label: __("Create Payment Entry"),
                    click(){
                        create_payment_entry_from_pi(voucher_no);
                        dialog.hide();
                    }
                }
            ]
        });
    } else if (voucher_type === "Purchase Receipt"){
        dialog = new frappe.ui.Dialog({
            title: __("Create for Purchase Receipt"),
            fields: [
                {
                    fieldtype: "Button",
                    label:__("Create Purchase Invoice"),
                    click(){
                        create_purchase_invoice_from_pr(voucher_no);
                        dialog.hide();
                    }
                }
            ]
        });
    } else {
        frappe.msgprint(__("No create actions for this voucher type"));
        return;
    }
    dialog.show();
}

File: demo/demo/report/voucher_action_report/voucher_action_report.py
import frappe
import json
from frappe.utils import flt, nowdate
from erpnext.controllers.accounts_controller import update_child_qty_rate

def execute(filters=None):  
    voucher_type = filters.get("voucher_type")
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")

    rows = get_sales_rows(voucher_type, from_date, to_date, filters)

    columns = [
        {"label": "Voucher Type", "fieldname": "voucher_type", "fieldtype": "Data", "width": 120},
        {"label": "Voucher No", "fieldname": "voucher_no", "fieldtype": "Link", "options": voucher_type, "width": 150},
        {"label": "Party", "fieldname": "party", "fieldtype": "Data", "width": 80},
        {"label": "Party Group", "fieldname": "party_group", "fieldtype": "Data", "width": 120},
        {"label": "Posting Date", "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 100},
        {"label": "Total Amount", "fieldname": "total_amount", "fieldtype": "Currency", "width": 120},
        {"label": "Outstanding Amount", "fieldname": "outstanding_amount", "fieldtype": "Currency", "width": 120},
        {"label": "Actions", "fieldname": "actions", "fieldtype": "Data", "width": 150},
        {"label": "View Items", "fieldname": "view_items", "fieldtype": "Data", "width": 100},
    ]
    
    for row in rows:
        row.update({
            "voucher_type": voucher_type,
            "outstanding_amount": row.get("outstanding_amount", 0),
            "actions": row.get("actions",""),
            "view_items": row.get("view_items","")
        })

    return columns, rows

def get_sales_rows(voucher_type, from_date, to_date, filters):
    date_fields = {
        "Sales Order": "transaction_date",
        "Sales Invoice": "posting_date",
        "Delivery Note": "posting_date",
        "Purchase Order": "transaction_date",
        "Purchase Invoice": "posting_date",
        "Purchase Receipt": "posting_date"
    }

    party_fields = {
        "Sales Order": ("customer", "customer_group", "Customer"),
        "Sales Invoice": ("customer", "customer_group", "Customer"),
        "Delivery Note": ("customer", "customer_group", "Customer"),
        "Purchase Order": ("supplier", "supplier_group", "Supplier"),
        "Purchase Invoice": ("supplier", "supplier_group", "Supplier"),
        "Purchase Receipt": ("supplier", "supplier_group", "Supplier")
    }

    outstanding_field_map = {
        "Sales Invoice": "outstanding_amount",
        "Purchase Invoice": "outstanding_amount"
    }

    alias_map = {
        "Sales Order": "so",
        "Sales Invoice": "si",
        "Delivery Note": "dn",
        "Purchase Order": "po",
        "Purchase Invoice": "pi",
        "Purchase Receipt": "pr"    
    }

    date_field = date_fields[voucher_type]
    party_field, party_group_field, party_table = party_fields[voucher_type]
    outstanding_field = outstanding_field_map.get(voucher_type, "0")
    alias = alias_map[voucher_type]

    conditions = [f"{alias}.docstatus < 2"]

    if voucher_type == "Sales Order":
        conditions.append(f"{alias}.per_delivered < 100")
        conditions.append(f"{alias}.per_billed < 100")
    elif voucher_type == "Purchase Order":
        conditions.append(f"{alias}.per_received < 100")
        conditions.append(f"{alias}.per_billed < 100")
    elif voucher_type in ["Sales Invoice", "Purchase Invoice"]:
        conditions.append(f"{alias}.outstanding_amount > 0")
    elif voucher_type in ["Delivery Note", "Purchase Receipt"]:
        conditions.append(f"{alias}.status != 'Completed'")

    values = {}

    if from_date:
        conditions.append(f"{alias}.{date_field} >= %(from_date)s")
        values["from_date"] = from_date
    if to_date:
        conditions.append(f"{alias}.{date_field} <= %(to_date)s")
        values["to_date"] = to_date
    if filters.get("status"):
        conditions.append(f"{alias}.status = %(status)s")
        values["status"] = filters["status"]
    if filters.get("customer") and party_field == "customer":
        conditions.append(f"{alias}.customer = %(customer)s")
        values["customer"] = filters["customer"]
    if filters.get("customer_group") and party_group_field == "customer_group":
        conditions.append(f"{party_table.lower()}.{party_group_field} = %(customer_group)s")
        values["customer_group"] = filters["customer_group"]
    if filters.get("supplier") and party_field == "supplier":
        conditions.append(f"{alias}.supplier = %(supplier)s")
        values["supplier"] = filters["supplier"]
    if filters.get("supplier_group") and party_group_field == "supplier_group":
        conditions.append(f"{party_table.lower()}.{party_group_field} = %(supplier_group)s")
        values["supplier_group"] = filters["supplier_group"]

    condition_str = " AND ".join(conditions)

    query = f"""
        SELECT
            {alias}.name AS voucher_no,
            {alias}.{party_field} AS party,
            {party_table.lower()}.{party_group_field} AS party_group,
            {alias}.{date_field} AS posting_date,
            {alias}.status,
            {alias}.rounded_total AS total_amount,
            {outstanding_field} AS outstanding_amount,
            '' AS actions,
            '' AS view_items
        FROM `tab{voucher_type}` {alias}
        LEFT JOIN `tab{party_table}` {party_table.lower()}
            ON {alias}.{party_field} = {party_table.lower()}.name
        WHERE {condition_str}
        ORDER BY {alias}.{date_field} DESC
    """

    return frappe.db.sql(query, values, as_dict=True)

@frappe.whitelist()
def get_items(voucher_type, voucher_no):
    if voucher_type == "Sales Order":
        items = frappe.get_all("Sales Order Item",
            filters={"parent": voucher_no},
            fields=["item_code", "item_name", "qty as total_qty", "rate", "amount", "delivered_qty"])
        for item in items:
            item["pending_qty"] = item["total_qty"] - item.get("delivered_qty", 0)
    elif voucher_type == "Sales Invoice":
        items = frappe.get_all("Sales Invoice Item",
            filters={"parent": voucher_no},
            fields=["item_code", "item_name", "qty", "rate", "amount"])
    elif voucher_type == "Delivery Note":
        items = frappe.get_all("Delivery Note Item",
            filters={"parent": voucher_no},
            fields=["item_code", "item_name", "qty", "rate", "amount"])
    elif voucher_type == "Purchase Order":
        items = frappe.get_all("Purchase Order Item",
            filters={"parent": voucher_no},
            fields=["item_code", "item_name", "qty", "rate", "amount", "received_qty"])
        for item in items:
            item["pending_qty"] = item["qty"] - item.get("received_qty", 0)
    elif voucher_type == "Purchase Invoice":
        items = frappe.get_all("Purchase Invoice Item",
            filters={"parent": voucher_no},
            fields=["item_code", "item_name", "qty", "rate", "amount"])
    elif voucher_type == "Purchase Receipt":
        items = frappe.get_all("Purchase Receipt Item",
            filters={"parent": voucher_no},
            fields=["item_code", "item_name", "qty", "rate", "amount"])
    return items

@frappe.whitelist()
def update_sales_purchase_order_items_custom(parent_doctype, trans_items, parent_doctype_name):
    update_child_qty_rate(parent_doctype, trans_items, parent_doctype_name, child_docname="items")
    
# Sales
@frappe.whitelist()
def create_combined_sales_invoice(sales_orders):
    if isinstance(sales_orders, str):
        sales_orders = json.loads(sales_orders)

    so_docs = [frappe.get_doc("Sales Order", so) for so in sales_orders]

    for so in so_docs:
        existing_si = frappe.db.exists({
            "doctype": "Sales Invoice Item",
            "sales_order": so.name,
            "parenttype": "Sales Invoice",
            "parentfield": "items"
        })
        if existing_si:
            frappe.throw(f"Sales Order {so.name} is already invoiced.")

    customers = list(set([so.customer for so in so_docs]))
    if len(customers) > 1:
        frappe.throw("Selected Sales Orders belong to different customers.")
    customer = customers[0]

    invoice = frappe.new_doc("Sales Invoice")
    invoice.customer = customer
    invoice.due_date = nowdate()
    invoice.naming_series = "ACC-SINV-.YYYY.-.#####"

    for so in so_docs:
        for item in so.items:
            pending_qty = flt(item.qty) - flt(item.delivered_qty)
            if pending_qty > 0:
                invoice.append("items", {
                    "item_code": item.item_code,
                    "qty": pending_qty,
                    "rate": item.rate,
                    "uom": item.uom,
                    "warehouse": item.warehouse,
                    "sales_order": so.name,
                    "description": item.description
                })

    if not invoice.items:
        frappe.throw("No pending items found to invoice.")

    invoice.flags.ignore_permissions = True
    invoice.insert()

    frappe.msgprint(f"Created Combined Sales Invoice <b>{invoice.name}</b> for Customer <b>{customer}</b>.")
    return invoice.name


@frappe.whitelist()
def create_combined_payment_entry(sales_invoices):
    if isinstance(sales_invoices, str):
        sales_invoices = json.loads(sales_invoices)

    si_docs = [frappe.get_doc("Sales Invoice", si) for si in sales_invoices]

    customers = list(set([si.customer for si in si_docs]))
    if len(customers) > 1:
        frappe.throw("Selected invoices belong to different customers.")
    customer = customers[0]

    unpaid_invoices = [si for si in si_docs if flt(si.outstanding_amount) > 0]
    if not unpaid_invoices:
        frappe.throw("All selected invoices are already fully paid.")

    pe = frappe.new_doc("Payment Entry")
    pe.payment_type = "Receive"
    pe.party_type = "Customer"
    pe.party = customer
    pe.posting_date = nowdate()
    pe.mode_of_payment = "Cash"
    pe.company = unpaid_invoices[0].company
    pe.paid_to = frappe.db.get_value("Company", pe.company, "default_cash_account")

    pe.received_amount = 0
    pe.paid_amount = 0

    for si in unpaid_invoices:
        amount = flt(si.outstanding_amount)
        pe.append("references", {
            "reference_doctype": "Sales Invoice",
            "reference_name": si.name,
            "due_date": si.due_date,
            "total_amount": si.grand_total,
            "outstanding_amount": si.outstanding_amount,
            "allocated_amount": amount
        })
        pe.received_amount += amount
        pe.paid_amount += amount

    pe.flags.ignore_permissions = True
    pe.insert()

    frappe.msgprint(f"Created Combined Payment Entry <b>{pe.name}</b> for Customer <b>{customer}</b>.")
    return pe.name

# Purchase
@frappe.whitelist()
def create_combined_purchase_invoice(purchase_orders):
    if isinstance(purchase_orders, str):
        purchase_orders = json.loads(purchase_orders)

    po_docs = [frappe.get_doc("Purchase Order", po) for po in purchase_orders]

    for po in po_docs:
        existing_pi = frappe.db.exists({
            "doctype": "Purchase Invoice Item",
            "purchase_order": po.name,
            "parenttype": "Purchase Invoice",
            "parentfield": "items"
        })
        if existing_pi:
            frappe.throw(f"Purchase Order {po.name} is already invoiced.")

    suppliers = list(set([po.supplier for po in po_docs]))
    if len(suppliers) > 1:
        frappe.throw("Selected Purchase Orders belong to different suppliers.")
    supplier = suppliers[0]

    invoice = frappe.new_doc("Purchase Invoice")
    invoice.supplier = supplier
    invoice.due_date = nowdate()
    invoice.naming_series = "ACC-PINV-.YYYY.-.#####"

    for po in po_docs:
        for item in po.items:
            pending_qty = flt(item.qty) - flt(item.received_qty)
            if pending_qty > 0:
                invoice.append("items", {
                    "item_code": item.item_code,
                    "qty": pending_qty,
                    "rate": item.rate,
                    "uom": item.uom,
                    "warehouse": item.warehouse,
                    "purchase_order": po.name,
                    "description": item.description
                })

    if not invoice.items:
        frappe.throw("No pending items found to bill.")

    invoice.flags.ignore_permissions = True
    invoice.insert()

    frappe.msgprint(f"Created Combined Purchase Invoice <b>{invoice.name}</b> for Supplier <b>{supplier}</b>.")
    return invoice.name


@frappe.whitelist()
def create_combined_purchase_payment_entry(purchase_invoices):
    if isinstance(purchase_invoices, str):
        purchase_invoices = json.loads(purchase_invoices)

    pi_docs = [frappe.get_doc("Purchase Invoice", pi) for pi in purchase_invoices]

    suppliers = list(set([pi.supplier for pi in pi_docs]))
    if len(suppliers) > 1:
        frappe.throw("Selected invoices belong to different suppliers.")
    supplier = suppliers[0]

    unpaid_invoices = [pi for pi in pi_docs if flt(pi.outstanding_amount) > 0]
    if not unpaid_invoices:
        frappe.throw("All selected invoices are already paid.")

    pe = frappe.new_doc("Payment Entry")
    pe.payment_type = "Pay"
    pe.party_type = "Supplier"
    pe.party = supplier
    pe.posting_date = nowdate()
    pe.mode_of_payment = "Cash"
    pe.company = unpaid_invoices[0].company
    pe.paid_from = frappe.db.get_value("Company", pe.company, "default_cash_account")

    pe.paid_amount = 0
    pe.received_amount = 0

    for pi in unpaid_invoices:
        amount = flt(pi.outstanding_amount)
        pe.append("references", {
            "reference_doctype": "Purchase Invoice",
            "reference_name": pi.name,
            "due_date": pi.due_date,
            "total_amount": pi.grand_total,
            "outstanding_amount": pi.outstanding_amount,
            "allocated_amount": min(amount, flt(pi.outstanding_amount, 2))
        })
        pe.paid_amount += amount
        pe.received_amount += amount

    pe.flags.ignore_permissions = True
    pe.insert()

    frappe.msgprint(f"Created Combined Payment Entry <b>{pe.name}</b> for Supplier <b>{supplier}</b>.")
    return pe.name
